test/metabase/api/session_test.clj

```diff
@@ -56,6 +58,73 @@
+                              "password" "incorrect-password"}
+                :content-type :json
+                :headers headers})
+    (catch clojure.lang.ExceptionInfo e
+      (:object (ex-data e)))))
+
+(defn- cleaned-throttlers [var-symbol ks]
+  (let [throttlers (var-get var-symbol)
+        clean-key  (fn [m k] (assoc-in m [k :attempts] (atom '())))]
+    (reduce clean-key throttlers ks)))
+
+;; Test that source based throttling kicks in after the login failure threshold (50) has been reached
+(expect
+  ["Too many attempts! You must wait 15 seconds before trying again."
+   "Too many attempts! You must wait 42 seconds before trying again."]
+  (with-redefs [session-api/login-throttlers          (cleaned-throttlers #'session-api/login-throttlers
+                                                                          [:username :ip-address])
+                public-settings/source-address-header (constantly "x-forwarded-for")]
+    (do
+      (dotimes [n 50]
+        (let [response    (send-login-request (format "user-%d" n)
+                                              {"x-forwarded-for" "10.1.2.3"})
+              status-code (:status response)]
+          (assert (= status-code 400) (str "Unexpected response status code:" status-code))))
+      [(-> (send-login-request "last-user" {"x-forwarded-for" "10.1.2.3"})
+            :body
+            json/parse-string
+            (get "errors")
+            (get "username"))
+       (-> (send-login-request "last-user" {"x-forwarded-for" "10.1.2.3"})
+            :body
+            json/parse-string
+            (get "errors")
+            (get "username"))])))
+
+;; The same as above, but ensure that throttling is done on a per request source basis.
+(expect
+  ["Too many attempts! You must wait 15 seconds before trying again."
+   "Too many attempts! You must wait 42 seconds before trying again."]
+  (with-redefs [session-api/login-throttlers          (cleaned-throttlers #'session-api/login-throttlers
+                                                                          [:username :ip-address])
+                public-settings/source-address-header (constantly "x-forwarded-for")]
+    (do
+      (dotimes [n 50]
+        (let [response    (send-login-request (format "user-%d" n)
+                                              {"x-forwarded-for" "10.1.2.3"})
+              status-code (:status response)]
+          (assert (= status-code 400) (str "Unexpected response status code:" status-code))))
+      (dotimes [n 50]
+        (let [response    (send-login-request (format "round2-user-%d" n)) ; no x-forwarded-for
+              status-code (:status response)]
+          (assert (= status-code 400) (str "Unexpected response status code:" status-code))))
+      [(-> (send-login-request "last-user" {"x-forwarded-for" "10.1.2.3"})
+            :body
+            json/parse-string
+            (get "errors")
+            (get "username"))
```

Comment:
Instead of multiple calls to `get` you can use [`get-in`](http://clojuredocs.org/clojure.core/get-in) e.g. `(get-in <whatever> ["errors" "username"])`